_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,347 @@
+
+- changed files
+- branches and remotes
+- commit list
+- stash
+- changes diff
+
+Using this interface, I will frequently:
+
+- build commits for a RFPR after a `git reset` to the base branch.
+- add individual lines across multiple files into a new or existing commit.
+- reorder commits on a feature branch.
+- amend a commit with a particular change.
+- take a change out of a commit and create a new commit from it.
+- ... and much more
+
+This post does not intend to be a tutorial on how to use `lazygit`. You should start with the
+[Tutorials](https://github.com/jesseduffield/lazygit?tab=readme-ov-file#tutorials) to get a sense
+for it. Getting a hang of the keybinds takes a short while, and then you can configure it to your
+liking.
+
+Believe me when I say that using a tool like this gives you a lot of power. So much so that when I
+create a RFPR, I spend more time on writing the documentation for it than actually building the
+commits using `lazygit` -- it's that good.
+
+## About Commit Mistakes
+
+The above approach is really helpful in ensuring that reviewers do not spend time giving feedback
+for changes that will never merge. Imagine you have a PR with 10 changes. The 5th commit introduces
+a bug which is then fixed in the 8th commit. Assuming that the reviewer will be reviewing in commit
+order, they will likely spend time giving feedback about the bug in the 5th commit, only to become
+irritated that they spent that time unnecessarily.
+
+Backing out all of your changes and then building the commits before the PR is created avoids this
+problem entirely. Performing a `git reset` to `origin/main` will in effect merge your changes
+together, only leaving the correct, bug-free diff.
+
+If after you've created your commits you notice a bug in the 5th commit, you definitely do not need
+to start all over. You can fix the bug in the working tree and then in `lazygit` navigate to the
+commit where the bug is. Hitting `A` will amend the selected commit with your changes. There are a
+lot of nice-to-haves like this in `lazygit`.
+
+# Receiving Feedback
+
+Once you've posted the RFPR and need to make followup commits, those commits should also be RFPR
+commits. Other people will come back later and read through them and we should also respect their
+time.
+
+I would highly recommend *not* rebasing or rewriting commits after you start getting feedback. The
+reason for this is that reordering the commits or amending them, ends up creating new commits under
+the hood (even if `lazygit` made it super easy to do so). Because reviewers will leave feedback on
+each commit during the initial review, if you rewrite that commit as a followup and force push, the
+feedback they left becomes orphaned and out of context because that commit no longer exists.
+
+For that reason, it's better to push new commits when addressing feedback, even if those changes
+would normally be structured differently if you were creating the RFPR from scratch.
```

Comment:
Something I just thought of reading this section. What's your stance on rebase+merge or squash-and-merge of PRs? Do you merge to transfer the commit history to `main` or do you squash it all together?

I'm a strong supporter of the squash-and-merge approach. The commits I build are meant for reviewers and I don't find them useful outside the context of a PR. When I `git blame` a line, I'm perfectly happy to open my browser and read the PR that created it. Not rewriting the commit history after people start reviewing means the commits at the end of a review may not be pristine.

Anyway, just a thought I had. It may be worthwhile to at least touch on the tradeoffs if you don't have a strong opinion about which method is "right".

Reply:
Thanks for bringing this up. I added a section at the end for this. The TLDR is that I don't feel strongly about it either way but I explained how I typically think about it. LMK what you think.